### PR TITLE
Validate user-provided `AuthSignature`

### PR DESCRIPTION
**Pull request by piotr-roslaniec**
**Type of PR:**
- Feature

**Required reviews:**
- 2

**What this does:**
- Implements `AuthSignature` validation with `zod`
- Moves some code around in `taco-auth` to prevent circular-dependency issues that break `zod`

**Issues fixed/closed:**
- Fixes https://github.com/nucypher/taco-web/issues/535

**Why it's needed:**
> Explain how this PR fits in the greater context of the NuCypher Network.
> E.g., if this PR address a `nucypher/productdev` issue, let reviewers know!

**Notes for reviewers:**
- ...

### CODE REVIEW

File: packages/taco-auth/src/storage.ts [hunk 1]
```diff
@@ -36,11 +38,24 @@ export class LocalStorage {
         : new BrowserStorage();
   }
 
-  getItem(key: string): string | null {
-    return this.storage.getItem(key);
+  public getAuthSignature(key: string): AuthSignature | null {
```
Comment (theref) [inline packages/taco-auth/src/storage.ts:41]: :rocket:

**Review by theref — APPROVED** [review]
Looks great

File: packages/test-utils/src/utils.ts [hunk 2]
```diff
@@ -53,8 +53,8 @@ export const fromBytes = (bytes: Uint8Array): string =>
 
 export const fakePorterUri = 'https://_this_should_crash.com/';
 
-const makeFakeProvider = (timestamp: number, blockNumber: number) => {
-  const block = { timestamp };
+const makeFakeProvider = (timestamp: number, blockNumber: number, blockHash: string) => {
```
Comment (manumonti) [inline packages/test-utils/src/utils.ts:56]: Not relevant for this PR, but shouldn't the code formatter split this line in several ones?


```suggestion
const makeFakeProvider = (
  timestamp: number,
  blockNumber: number,
  blockHash: string,
) => {
```

File: packages/shared/src/schemas.ts [hunk 3]
```diff
@@ -0,0 +1,4 @@
+import { z } from 'zod';
+
+export const ETH_ADDRESS_REGEXP = new RegExp('^0x[a-fA-F0-9]{40}$');
+export const EthAddressSchema = z.string().regex(ETH_ADDRESS_REGEXP);
```
Comment (cygnusv) [inline packages/shared/src/schemas.ts:4]: Should we go a step further and validate the address with ethers and a zod `.refine`?

https://docs.ethers.org/v6/api/address/#getAddress

File: packages/taco-auth/src/storage.ts [hunk 4]
```diff
@@ -36,11 +38,24 @@ export class LocalStorage {
...
+    this.storage.setItem(key, asJson);
+  }
+
+  public static fromJson(json: string): AuthSignature {
+    return authSignatureSchema.parse(JSON.parse(json));
   }
 
-  setItem(key: string, value: string): void {
-    this.storage.setItem(key, value);
+  public static toJson(signature: AuthSignature): string {
```
Comment (cygnusv) [inline packages/taco-auth/src/storage.ts:58]: Not a blocker but it's a bit weird that json<=>authsignature conversion methods are part of the storage interface

**Review by manumonti — APPROVED** [review]
LGTM! 👌

File: packages/taco-auth/src/providers/eip4361.ts [hunk 5]
```diff
@@ -1,10 +1,13 @@
...
+import { z } from 'zod';
 
+import { EIP4361_AUTH_METHOD } from '../auth-provider';
+import { AuthSignature } from '../auth-sig';
 import { LocalStorage } from '../storage';
-import { AuthSignature, EIP4361_AUTH_METHOD } from '../types';
 
-export type EIP4361TypedData = string;
+
+export const EIP4361TypedDataSchema = z.string();
```
Comment (derekpierre) [inline packages/taco-auth/src/providers/eip4361.ts:10]: Should we use zod `refine` here as well? You can try calling the `SiweMessage` constructor as part of the refine function to ensure that the string can be parsed properly.

Comment (piotr-roslaniec) [reply packages/taco-auth/src/providers/eip4361.ts:10]: Addressed here: [`f9e2030` (#527)](https://github.com/nucypher/taco-web/pull/527/commits/f9e203023327dc343e9aaf26fe2a362d5c8f2683)

Comment (derekpierre) [reply packages/taco-auth/src/providers/eip4361.ts:10]: 🎸

**Pull request merged**